base/expr.jl

```diff
@@ -570,6 +571,19 @@ moved between tasks without observable results.
     code that is not `:notaskstate`, but is `:effect_free` and `:consistent`
     may still be dead-code-eliminated and thus promoted to `:total`.
 
+---
+## `:noglobal`
```

Comment:
Is this the same as `inaccessiblemem_or_argmemonly` in [llvm](https://llvm.org/docs/LangRef.html#function-attributes)? Asking because at some point it would be useful to start transferring our annotations into LLVM too

Reply:
I don't think it's quite the same, since this excludes the case where you're passing in a global as an argument. I think `:noglobal` probably implies it though?

Reply:
Currently `:noglobal === true` means `inaccessiblemem_or_argmemonly` and it includes cases we access to (potentially) global memory passed as an argument. And we additionally check that arguments are never be mutable global memory (by a simple type analysis) when trying to refine `:consistent` or `:effect_free` information later.

Now I wonder if we should make `:noglobal === ALWAYS_TRUE` mean `inaccessiblemem` and `:noglobal === INACCESSIBLEMEM_OR_ARGMEMONLY` mean that so.

And I guess one more difference is that `:noglobal` doesn't really reason about immutable global memory, as it doesn't taint `:consistent`-cy/`:effect_free`-ness anyway.

Reply:
After thinking about this for a bit, now I think it's a bit confusing that currently `:noglobal === ALWAYS_TRUE` doesn't exclude access to arguments memory, and we may want to make `:noglobal === ALWAYS_TRUE` mean `inaccessiblemem` and `:noglobal === ARGMEM` mean `inaccessiblemem_or_argmemonly` and also rename `:noglobal` to `:inaccessiblemem`.

This property may not necessarily 1-to-1 correspond to LLVM's function attributes since our `:inaccessiblemem` only reasons about mutable memory.

Reply:
llvm also only reasons about mutable memory